### ISSUE TITLE
Migration scripts for v9

### ISSUE THREAD

**Issue by pedrobaeza** [OCA/OpenUpgrade#400]
This issue summarizes the development of the migration scripts for v9. Please write down a comment if you want to develop a migration script for one module, to avoid duplicating efforts between the contributors. I have extracted the module list to have it as a reference for tracking the progress:
- [x] base - Done by @coleste, @hbrunn 
- [x] account - Partial covered by @coleste, completed by @hbrunn
- [x] account_accountant - Nothing to do
- [x] account_analytic_default - Nothing to do
- [x] account_asset - Done by @hbrunn 
- [x] account_budget - Nothing to do
- [x] account_cancel - Nothing to do
- [x] account_chart (`del`) - By @cubells - #1070 
- [x] account_check_writing > account_check_printing - Done by @hbrunn
- [ ] account_tax_cash_basis
- [ ] account_tax_python
- [ ] account_test
- [x] account_voucher - Assigned to @jbeficent
- [x] analytic - Assigned to @hbrunn 
- [ ] anonymization
- [x] association - By @pedrobaeza - #911 
- [x] auth_crypt - Nothing to do
- [x] auth_ldap - Nothing to do
- [x] auth_oauth
- [x] auth_signup - Done by @hbrunn
- [x] barcodes - Nothing to do
- [x] base_action_rule - Nothing to do
- [ ] base_gengo
- [ ] base_geolocalize
- [x] base_iban - Nothing to do
- [x] base_import - Nothing to do
- [x] base_import_module - Nothing to do
- [x] base_setup - Nothing to do
- [x] base_vat
- [x] board - Nothing to do
- [x] bus - Done by @hbrunn
- [x] calendar - Nothing to do
- [x] claim_from_delivery - Done by @hbrunn
- [x] crm - Completed by @hbrunn , @coleste 
- [x] crm_claim - Done by @hbrunn 
- [ ] crm_partner_assign
- [ ] crm_project_issue
- [x] decimal_precision - Nothing to do
- [x] delivery - Done by @hbrunn 
- [x] document - Completed by @hbrunn 
- [x] event - By @cubells - #712 
- [x] event_sale
- [x] fetchmail - Nothing to do
- [x] fleet - By @cubells - #831 
- [x] gamification - By @cubells - #830 
- [x] gamification_sale_crm - By @cubells - #832 
- [x] google_account
- [x] google_calendar
- [x] google_drive
- [x] google_spreadsheet
- [x] hr - Done by @coleste, @hbrunn
- [x] hr_attendance -  Done by @hbrunn
- [x] hr_contract - Done by @hbrunn
- [ ] hr_equipment
- [x] hr_expense - assigned to @jbeficent
- [x] hr_gamification
- [x] hr_holidays - Done by @hbrunn
- [x] hr_payroll - Assigned to @jbeficent
- [x] hr_payroll_account - Assigned to @jbeficent
- [x] hr_recruitment - by @cubells #682 
- [x] hr_timesheet - assigned to @jbeficent
- [x] hr_timesheet_sheet - assigned to @jbeficent
- [x] hr_timesheet_invoice - obsolete. Create sale orders for any remaining invoicable lines
- [ ] hw_escpos
- [ ] hw_posbox_homepage
- [ ] hw_posbox_upgrade
- [ ] hw_proxy
- [ ] hw_scale
- [ ] hw_scanner
- [x] im_livechat - By @NL66278 - #737 
- [ ] im_odoo_support
- [x] l10n_fr - By @cubells - #1068 
- [x] l10n_nl - By @StefanRijnhart - #1203 
- [ ] link_tracker
- [ ] lunch
- [x] mail - Completed by @hbrunn, @coleste
- [ ] mail_tip
- [x] marketing - Nothing to do
- [x] marketing_campaign - By @pedrobaeza - #708
- [ ] marketing_campaign_crm_demo
- [x] mass_mailing - By @cubells - #677 
- [x] membership - By @mreficent - #919
- [x] mrp - Assigned to @jbeficent
- [x] mrp_byproduct - Nothing to do
- [x] mrp_operations - Nothing to do
- [x] mrp_repair - Nothing to do
- [x] note - Nothing to do
- [ ] note_pad
- [ ] pad
- [ ] pad_project
- [x] payment  - By @cubells - #675 
- [ ] payment_adyen
- [ ] payment_authorize
- [ ] payment_buckaroo
- [ ] payment_ogone
- [x] payment_paypal - By @pedrobaeza - #719 
- [ ] payment_sips
- [x] payment_transfer
- [ ] point_of_sale - By @sergio-teruel - #694
- [x] portal - By @cubells - #1029
- [x] portal_claim (deleted) - By @cubells - #1021 
- [ ] portal_gamification
- [ ] portal_sale
- [x] portal_stock - By @cubells - #1018 
- [ ] pos_cache
- [ ] pos_discount
- [ ] pos_mercury
- [ ] pos_reprint
- [x] pos_restaurant - By @legalsylvain - #1645 
- [x] procurement - By @cubells - #672 
- [x] procurement_jit
- [x] procurement_jit_stock - By @pedrobaeza - #927
- [x] product - Completed by @hbrunn 
- [ ] product_email_template
- [ ] product_expiry
- [ ] product_extended
- [ ] product_margin
- [ ] product_uos
- [x] product_visible_discount
- [x] project - Completed by @hbrunn 
- [x] project_issue - Done by @hbrunn, @vincentcoll 
- [x] project_issue_sheet - Assigned to @hbrunn 
- [x] project_timesheet
- [x] purchase - Assigned to @StefanRijnhart 
- [x] purchase_double_validation (`del`) - By @pedrobaeza - #2137 
- [x] purchase_requisition - Assigned to @StefanRijnhart 
- [ ] rating
- [ ] rating_project
- [ ] rating_project_issue
- [x] report - Nothing to do
- [ ] report_intrastat
- [x] report_webkit - Nothing to do
- [x] resource - Nothing to do
- [x] sale - Completed by @coleste + @StefanRijnhart 
- [x] sale_crm - Assigned to @StefanRijnhart #625
- [ ] sale_expense
- [ ] sale_layout
- [x] sale_margin - Nothing to do
- [x] sale_mrp - Nothing to do - By @mreficent - #913 
- [x] sale_order_dates - Nothing to do
- [x] sale_service - By @cubells - #833 
- [x] sales_team - Completed by @StefanRijnhart 
- [x] sale_stock - Completed by @StefanRijnhart
- [ ] sale_timesheet
- [x] stock - Completed by @hbrunn 
- [x] stock_account - Done by @hbrunn 
- [ ] stock_calendar
- [x] stock_dropshipping
- [x] stock_landed_costs - By @pedrobaeza - #2971 
- [x] stock_picking_wave - By @cubells - #1016 
- [x] subscription - Nothing to do
- [x] survey - By @cubells - #674 
- [x] survey_crm
- [ ] theme_bootswatch
- [ ] theme_default
- [x] utm - Nothing to do
- [x] warning - Nothing to do
- [x] web - Nothing to do
- [ ] web_analytics
- [x] web_calendar - Nothing to do
- [x] web_diagram - Nothing to do
- [ ] web_editor
- [x] web_kanban - Nothing to do
- [x] web_kanban_gauge
- [x] web_planner - Nothing to do
- [ ] web_settings_dashboard
- [x] website - By @pedrobaeza - #711 
- [x] website_blog - By @pedrobaeza - #720 
- [x] website_crm
- [x] website_crm_claim - By @cubells - #1021 
- [ ] website_crm_partner_assign
- [ ] website_customer
- [x] website_event
- [ ] website_event_questions
- [x] website_event_sale
- [x] website_event_track - By @pedrobaeza - #721 
- [ ] website_form
- [ ] website_forum
- [ ] website_forum_doc
- [ ] website_gengo
- [ ] website_google_map
- [ ] website_hr
- [ ] website_hr_recruitment
- [ ] website_issue
- [ ] website_links
- [ ] website_livechat
- [x] website_mail
- [ ] website_mail_channel
- [ ] website_mass_mailing
- [ ] website_membership
- [x] website_partner
- [x] website_payment - By @cubells - #1019 
- [ ] website_portal
- [ ] website_portal_sale
- [ ] website_project_issue
- [ ] website_project_issue_sheet
- [ ] website_quote
- [ ] website_rating_project_issue
- [x] website_report - By @pedrobaeza - #722 
- [x] website_sale - By @cubells - #676 
- [ ] website_sale_delivery
- [ ] website_sale_digital
- [ ] website_sale_options
- [ ] website_sale_stock
- [ ] website_twitter
- [x] web_tip - Nothing to do
- [ ] web_view_editor

OCA modules
----------

- [x] OCA/bank-statement-reconcile: bank_statement_period_from_line_date - Merged in account - By @cubells - #1071 
- [x] OCA/stock-logistics-workflow: stock_dropshipping_dual_invoice - Merged in stock_dropshipping - By @cubells - #1082 
- [x] OCA/sale-workflow: account_invoice_reorder_lines - Included in core - By @pedrobaeza - #1210 
- [x] OCA/product-attribute: product_pricelist_fixed_price - Merged in product - By @pedrobaeza - #1471

**Comment by Yenthe666** [comment]
@pedrobaeza and @StefanRijnhart I soon need a migration script from V8 to V9 so I was wondering how I could help and how this works. Note that my database knowledge is very limited but I would love to help in this. Could you guys give me information about how I should do this?

**Comment by pedrobaeza** [comment]
Sorry, but this is not very easy to achieve. There's a lot of work to do. You have to at least check that there are no changes on each of the modules you have installed in your v8 DB, and the ones that have changed their data layout, provide the corresponding scripts that transforms and manipulate data to adapt to the new layout. You have a bunch of examples in the migration scripts from previous versions, and basically is to programatically perform that changes via SQL, ORM or whatever you need. Check the available documentation: https://doc.therp.nl/openupgrade

**Comment by dreispt** [comment]
I'm in a similar position to @Yenthe666 , but I think the `base` module is the place to start.

**Comment by coleste** [comment]
I started working on base

**Comment by pedrobaeza** [comment]
Cool! @StefanRijnhart, did you start also with that task for not duplicating efforts?

**Comment by ghost** [comment]
I started working on Project module today.

**Comment by pedrobaeza** [comment]
I assign you then on the main text. Tell us any doubt about it, but please open a new issue for specific things on a module migration.

**Comment by ghost** [comment]
Understood, thanks!

**Comment by dreispt** [comment]
@coleste @mel-microcom You're welcome to open a [WIP] PR, so that we can follow progress.

**Comment by coleste** [comment]
Done in #404 but there are bootstrap issues I think.

**Comment by mmalorni** [comment]
@pedrobaeza @StefanRijnhart please have a look at PR #405  and #407  as they are needed to get #404 to pass the tests.
Anyone else want to review?
Thanks,

**Comment by luanads** [comment]
Hi guys, what do you think about using zenhub.io to manage the migration scrips in different issues on a kanban board?

**Comment by mileo** [comment]
@luanads Take a look at: https://waffle.io/OCA/OpenUpgrade

**Comment by okuryan** [comment]
Hi @pedrobaeza, Is there an overview of what is done for migration from v8 to v9 and what is pending? 
Is list of modules in first message is actual?

**Comment by pedrobaeza** [comment]
Yeah, that's it. In summary, there's a lot to do.

**Comment by pedrobaeza** [comment]
Some notes for making _account_ module migration:
- `amount` field is multiplied by 100 in this version. 8.0: `0.21`. 9.0: `21`
- column rename in account.tax: `type` > `amount_type`
- column rename in account.account: `user_type` > `user_type_id`
- In v8, if `child_depend == True`, then in v9, `amount_type='group'`

**Comment by esousy** [comment]
do you have any idea how to migrate payments and reconciliations?

Regards

**Comment by pedrobaeza** [comment]
Not related to this project, so I remove both comments.

**Comment by mmalorni** [comment]
@pedrobaeza , @StefanRijnhart  ,  . @coleste is working on the CRM, HR, Sales, Account, project, project issue, mail, product and stock. We might not finish the job but we did the changes we needed to get the OpenUpgrade passing with our data for those modules. We will do pull requests once we have finished the tests and have fixed the bug. Thought you might want to indicate that someone is working on those in the list above.

**Comment by pedrobaeza** [comment]
Written in the main comment, but it's a pity to let a half work. Please reconsider to give a bit extra fuel to complete that scripts.

**Comment by mmalorni** [comment]
@pedrobaeza,  @StefanRijnhart, .@coleste has done a number of pull requests for the modules CRM, HR, Sales, Account, project, project issue, mail, product and stock. Those work for our environment but we feel that some features or fields might be used by others which will not be covered by the code. 
The parts that have not been done should be identified in the working document of the analysis for each module. We are available to answer questions and fix issues in the existing scripts or to help someone continue the work on the different modules. We do not have the time right now to complete them all.
Some issues might be easy to fix once the analysis is done but our understanding of the differences is limited to the analysis and it is not always sufficient. For example, we realized that the flag to identify unread messages has changed in the new mail object but we didn't investigate it and didn't do a mapping between the old system and the new. The result is that after converting the messages, they are all marked as read. It wasn't a big issue so we did not dig deeper. If someone lets us know, we might take the time to fix while we have the environment setup.

**Comment by mmalorni** [comment]
@pedrobaeza, @StefanRijnhart,@dreispt
 We did a migration of our own server with those modules and the only limitation was that the mail history was lost. Other limitations are listed in the analysis document of each module.
Some could potentially be reviewed and merged.

[WIP] 9.0 - stock #494 opened 15 days ago by @coleste (Anyone can get the PR and update it)
[WIP] 9.0 - sales #493 opened 15 days ago by @coleste  (Anyone can get the PR and update it)
[WIP] 9.0 mail #492 opened 15 days ago by @coleste  (Anyone can get the PR and update it)
[WIP] hr - remove view_users_form_mail #491 opened 15 days ago by @coleste  (Anyone can get the PR and update it)
[WIP] 9.0 crm #490 opened 15 days ago by @coleste  (Anyone can get the PR and update it)
[FIX] [Review] 9.0 base #489 opened 15 days ago by @coleste 
[WIP] 9.0 account #488 opened 15 days ago by @coleste

**Comment by jholze** [comment]
@StefanRijnhart @pedrobaeza We are in planning to migrate to the odoo version 10 in the Q1 of 2017 and the plan is to use and support the openupgrade project. What do you guys think will be the best way, to first finish the migration scripts for 9 and than update / add new for 10, or to write the scripts for 10 and downgrade the specific scripts to 9 ?

**Comment by hbrunn** [comment]
This depends on what you want to upgrade. If it's a v9 database, the scripts migrating to v9 are not interesting for you. If you want to migrate a v8 or lower database to 10, you'll need the v9 scripts anyways (openupgrade is always incremental)

You might want to talk to @jbeficent who gathered the funds for a big part of the current v9 migration to check if you allocated enough resources.

**Comment by jholze** [comment]
ok, thanks for both. I didn't knew that the upgrade scripts are incremental.

**Comment by roberto-barreiro** [comment]
Is this list updated? Purchase module isn't still supported for migration?

**Comment by pedrobaeza** [comment]
Indeed that one was forgotten to be update. Done on https://github.com/OCA/OpenUpgrade/commit/1f5b73acb143465d9e0ce30f1d6d54dd8caefe65

**Comment by roberto-barreiro** [comment]
Thank you for answering and updating Pedro.

**Comment by jholze** [comment]
@pedrobaeza i have good news, we got a project to migrate from odoo 7 to 10 and we will start asap to support the openupgrade with our capacity. I hope we can help to finish some of the migrations scripts which are still open.

**Comment by pedrobaeza** [comment]
Cool, Jens!

First task is to compare your installed modules in the v7 with the script coverage and find the gaps for migrating on each version.

Then, you can inform here which modules are you going to add scripts for, and we can assign you.

When the PRs are done, I can review them.

**Comment by jholze** [comment]
Cool. Our devs @rruebner @ateijelo will start soon with this. Just a question, is there a similar tickets also vor v10 ?

**Comment by pedrobaeza** [comment]
Not for now, but the summary is that everything is pending :disappointed: 

I'm talking also with @jbeficent, that is interested in join efforts. He is going to send us soon an email for coordinating the effort.

**Comment by jholze** [comment]
Ok thanks. Another good news are, that we are also on a second very very big project where we also  in discussion with the customer to migrate this as well. This is 8 to 10. so we should create a checklist for the 9 to 10 migration soon.

**Comment by rruebner** [comment]
Hi all,

@pedrobaeza, after checking our modules I have a few questions to some of them.

We use some modules which were removed in odoo v9. For some modules we found alternatives (OCA, EE, etc.) but there are some modules without alternatives. What should we do with them? Just ignore? 
* account_analytic_analysis
* account_chart
* edi (I know that the edi handling was done in each module separately before)
* email_template (is the migration done in the mail module, like move the model email.template from module email_template to mail.template in module mail)
* hr_timesheet_invoice
* share
* web_graph
* web_tests

Do we need to migrate the l10n_XX modules? I ask because I don't see them on the list

If nobody has taken base_vat and web_view_editor so far we would do the migration for them.

@jbeficent  Can you say how is the migration progress for account_voucher and hr_timesheet.
@hbrunn  Can you say how is the migration progress for analytic.
@cubells Can you say how is the migration progress for procurement.
@StefanRijnhart Can you say how is the migration progress for purchase.

**Comment by JordiBForgeFlow** [comment]
@rruebner what do you mean by 'how is the migration'? Can you clarify?

**Comment by pedrobaeza** [comment]
> account_analytic_analysis

This is already handled on contract module for OCA/contract. Module renamings are handled in https://github.com/OCA/OpenUpgrade/blob/9.0/openerp/addons/openupgrade_records/lib/apriori.py, and once renamed at the beginning of the migration process, you can add pre and post migration script if your target module, like it's done on https://github.com/OCA/contract/blob/9.0/contract/migrations/9.0.1.0.0/pre-migration.py.

> account_chart

I don't know about this one and what is their function. The technique here is to see if the functionality is included on core already. If so, look at the explanation of _email_template_. If not, then you can migrate the module and put in any OCA repository. If you change the name of the module because of any requirement (meet OCA conventions, avoid conflicts with other existing modules...), you know already where to do this renaming :wink: 

> edi (I know that the edi handling was done in each module separately before)

The functionality of this module has disappeared, and for now the decision has been to not supply an alternative, as nobody uses it, but as this is a required module in v8 due to dependency chain, the used approach is the same as the _email_template_ for avoiding later actions by the user.

> email_template (is the migration done in the mail module, like move the model email.template from module email_template to mail.template in module mail)

The functionality of this module has been merged on _mail_ module, so it was handled in the mail module migration (as that one is a dependency) with a call to an openupgradelib method that allows that thanks to an expansion I did in PR https://github.com/OCA/openupgradelib/pull/63: https://github.com/OCA/OpenUpgrade/blob/8120ca0ac64ed6c883f0e009c92f9a0d2855b3fc/addons/mail/migrations/9.0.1.0/pre-migration.py#L25.

If anything more needs to be done, it was treated in the same migration script (for example, column renames).

> hr_timesheet_invoice

This module is also merged with contract main functionality in the contract migration: https://github.com/OCA/contract/blob/191d690d9669a6a33f35bdc83ef72f0a2940f61c/contract/migrations/9.0.1.0.0/pre-migration.py#L17

> share

This function has also disappeared, but for now no decision has been made about easing migration merging its definition on other module (probably base). I think we should merge it indeed. If not merged, it's uninstalled on v9 and it's also a no brainer.

> web_graph
> web_tests

Both are merged on _web_ main module: https://github.com/OCA/OpenUpgrade/blob/8120ca0ac64ed6c883f0e009c92f9a0d2855b3fc/openerp/addons/base/migrations/9.0.1.3/pre-migration.py#L73.

**Comment by rruebner** [comment]
@pedrobaeza thanks a lot for the information.

@jbeficent We just want to know how far you are with the migration scripts for account_voucher and hr_timesheet. Because we need to migrate these modules too in our project. So if you are not done with the migration scripts maybe we can participate to finish the migration scripts.

**Comment by JordiBForgeFlow** [comment]
@rruebner we're working on both.

**Comment by ateijelo** [comment]
Hi, @jbeficent . How is progress coming along on account_voucher ? How can I help you?

**Comment by JordiBForgeFlow** [comment]
Should be done, both voucher and hr_timesheet.

**Comment by ateijelo** [comment]
Thanks @jbeficent . I still offer my help if it's needed.
@hbrunn, you said `analytic` is done "tentatively". What do you mean?
I could assist in testing or improving the migration.

**Comment by hbrunn** [comment]
back then I wrote those scripts in a rush, and as soon as it was good enough for me, I released it with the marker that there might be something else to do for other databases. But now after a couple of migrations I think we'd have found problems, so you can make a PR to change this to just done if it also works for you

**Comment by mlaitinen** [comment]
I just compared branches `8.0` and `9.0`. `auth_ldap` doesn't have any code changes. 

Also, `rating`, `rating_project` and `rating_project_issue` don't exist in Odoo 8.

IMO all of these should be marked as "Nothing to do".

**Comment by Martronic-SA** [comment]
@pedrobaeza 

> hr_timesheet_invoice

>This module is also merged with contract main functionality in the contract migration: https://github.com/OCA/contract/blob/191d690d9669a6a33f35bdc83ef72f0a2940f61c/contract/migrations/9.0.1.0.0/pre-migration.py#L17

I don't see a real link between hr_timesheet_invoice and contract. Contracts are made to produce recurring invoices and hr_timesheet_invoice is supposed to make invoices from account analytic lines (tasks and material to invoice).

**Comment by StefanRijnhart** [comment]
@mlaitinen thank you for checking these modules. We mark modules 'Nothing to do' by filing a pull request. Will you do the honours?

**Comment by Martronic-SA** [comment]
> hr_timesheet_invoice
This module is also merged with contract main functionality in the contract migration: https://github.com/OCA/contract/blob/191d690d9669a6a33f35bdc83ef72f0a2940f61c/contract/migrations/9.0.1.0.0/pre-migration.py#L17

Ok if I understand well, the hr_timesheet (hr.analytic.timesheet model) has been removed in profit of pure analytic line. Why not as the hr.analytic.timesheet was "just" a renamed analytic line.
So we are missing a good community module to make invoices from analytic lines that would replace the disappeared hr_timesheet_invoice.

**Comment by mlaitinen** [comment]
@StefanRijnhart I had a hunch this was the wrong way to do it :) Thanks for leading me to the right direction.

I just created PR #776.

**Comment by igallart** [comment]
Migrate module note fail, the field group_note_fancy not is present in odoo 9.

**Comment by pedrobaeza** [comment]
That's not generally a problem that a column dissapears.

**Comment by pepetreshere** [comment]
Hello everybody.  

At Trescloud we have started a migration project, we need to get with a set of modules up to v10. 

I understand we must first ensure everything is migrable to v9, so would somebody please help me clarify if the following modules are ready to be migrated to v9?

- procurement_jit_stock -- Which is supossed to have been merged into procurement_jit since v9
- stock_landed_costs
- base -- I don't see the module in the list... but it is core.
- account_anglo_saxon -- although we might remove this module
- account_chart
- knowledge
- edi (which should be removed as no longer used, but is not listed in the modules80-90.rst file)
- email_template
- portal

**Comment by hbrunn** [comment]
it's not clear from [the documentation](https://github.com/OCA/OpenUpgrade/blob/9.0/openerp/openupgrade/doc/source/modules80-90.rst) that `email_template` has been ported because this functionality is provided by the `mail` template by now. The `base` module is ported, that's implicit as nothing works without this. Some other modules just lack an entry in the list, just check the `migrations/$version` folder. If this contains a file `openupgrade_analysis_work.txt`, somebody looked at at already, then it either says there's nothing to do or what needs to be done. `account_anglo_saxon` is a configuration in `account` by now, `account_chart` too

**Comment by pepetreshere** [comment]
Thanks @hbrunn for your quick reply. I am trying to start learning and working on this project.

I most likely will support the migration to version 10 as it seems that all the modules we used in version 8 are already migrated to version 9 :) (module list might need update however). I have will be doing tests during this week however and I will come back then for picking a module for migration.

**Comment by sixtyfive** [comment]
hi all, am i correct to understand that migrating from odoo 8 to odoo 9 is still not possible without paying money to The Company? the above task list looks a little like it might depend on the database. is there a script somewhere to check an existing odoo 8 database for whether it's upgradable?

**Comment by pedrobaeza** [comment]
@sixtyfive it's very simple to check by yourself comparing installed modules with this list (or more specifically, the one included in documentation). Current migration scripts cover 99% of Odoo installations. Any way, this requires a deep knowledge of Odoo for solving inconsistencies in your data, so you will might need to pay "The Company?" or other integrator if you want to be sure a good and professional work is done migrating your DB.

**Comment by pedrobaeza** [comment]
I think we can close this as most relevant modules have been covered, and if someone comes with a missing one, it's not needed to be opened to be tracked here.

If you don't agree, please comment back.
